File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        Map<Node,Node> clones = new HashMap<>();
        return clone(node,clones);
    }
    public Node clone(Node node, Map<Node,Node> clones){
        if (null==node) return null;

        if (clones.containsKey(node)) return clones.get(node);

        Node copy = new Node(node.val);
        clones.put(node,copy);
        for(Node n:node.neighbors){
            copy.neighbors.add(clone(n,clones));
        }
        return copy;
    }
}

File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, root: Optional['Node']) -> Optional['Node']:
        clones = dict()

        def clone(node):
            copy = None
            if node:
                if node.val not in clones:
                    copy = Node(node.val)
                    clones[node.val] = copy
                    for n in node.neighbors:
                        copy.neighbors.append(clone(n))
                else:
                    copy = clones[node.val]
            return copy
        return clone(root)
                    
